rllib/evaluation/postprocessing.py

```diff
@@ -16,6 +16,55 @@ class Postprocessing:
     VALUE_TARGETS = "value_targets"
 
 
+def adjust_nstep(n_step: int, gamma: float, batch: SampleBatch) -> None:
+    """Rewrites `batch` to encode n-step rewards, dones, and next-obs.
+
+    Observations and actions remain unaffected. At the end of the trajectory,
+    n is truncated to fit in the traj length.
+
+    Args:
+        n_step (int): The number of steps to look ahead and adjust.
+        gamma (float): The discount factor.
+        batch (SampleBatch): The SampleBatch to adjust (in place).
+
+    Examples:
+        n-step=3
+        Trajectory=o0 r0 d0, o1 r1 d1, o2 r2 d2, o3 r3 d3, o4 r4 d4=True o5
+        gamma=0.9
+        Returned trajectory:
+        0: o0 [r0 + 0.9*r1 + 0.9^2*r2 + 0.9^3*r3] d3 o0'=o3
+        1: o1 [r1 + 0.9*r2 + 0.9^2*r3 + 0.9^3*r4] d4 o1'=o4
+        2: o2 [r2 + 0.9*r3 + 0.9^2*r4] d4 o1'=o5
+        3: o3 [r3 + 0.9*r4] d4 o3'=o5
+        4: o4 r4 d4 o4'=o5
+    """
+
+    assert not any(batch[SampleBatch.DONES][:-1]), \
```

Comment:
just to double check, I remember reading about a "truncated_episodes" batch mode, where the collected batch may span across multiple batches.
will that cause problem here?

Reply:
No, because we only check up till the last DONE (excluding the last one). Making sure that none of the other dones are True -> We check, whether all data in the trajectory is from the same episode (but this episode does not have to be completed yet).
If the episode is not done, the estimate will be somewhat "crunched" (b/c we will have n-step=1/2/3 at the end of the chunk for n-step=3), but that's ok. For n-step learning, it's ok to actually mix the step size within a batch. We should actually consider this: Mix different n-steps within the same run, similar to how TD(lambda) does it.

Reply:
totally understand. wondering if we should fail on the case when n-step>1 and batch_mode=truncated_episode during config validation?

Reply:
No, we shouldn't. It's totally valid to have n-step = 3, but then - for truncated episodes - have a few n-step=1|2 tuples (where the truncation is) in the data.

Reply:
oh sven, to clarify, I meant this chart from our documentation:
https://docs.ray.io/en/latest/_images/rllib-batch-modes.svg
you see that for truncate_episode mode, there may be DONEs in the middle of the batch.

Reply:
True, in the final train-batch, but not for the call to postprocess function, which get called as soon as one rollout-fragment is complete (either episode ends or the rollout fragment lengths is reached). So inside any postprocess function, there should always only be one episode. We should actually have a more central assert in the Sampler to clarify this for the users (instead of doing asserts in the different postprocess functions).

Reply:
ah, ok, now I understand!! this makes total sense. thanks for your patience Sven :)